### PR TITLE
ActionController::Parameters#deep_dup

### PR DESCRIPTION
**Pull request by Paxa**
Provide `params.deep_dup` which actually create copy

Fixes https://github.com/rails/rails/issues/26566

### CODE REVIEW

**Comment by Paxa** [comment]
Should it have some tests?

**Comment by Paxa** [comment]
Any thoughts?

**Comment by arthurnn** [comment]
hi @Paxa , thanks for the PR.
Can you add some tests for it.
I think it looks good to have a deep_dup method in here.
Let me know once you added the tests, and I will take a final look.
Thanks

**Comment by Paxa** [comment]
@arthurnn Tests ready, please take a look

File: actionpack/test/controller/parameters/dup_test.rb [hunk 1]
```diff
@@ -40,4 +41,15 @@ class ParametersDupTest < ActiveSupport::TestCase
...
+
+  test "deep_dup" do
+    dupped_params = @params.deep_dup
+    dupped_params[:person][:age] = "45"
+    dupped_params[:person][:addresses].clear
+    dupped_params.permit!
+
+    assert_not @params.permitted?
+    assert_not_equal @params[:person][:age], dupped_params[:person][:age]
+    assert_not_equal @params[:person][:addresses], dupped_params[:person][:addresses]
```
Comment (rafaelfranca) [inline actionpack/test/controller/parameters/dup_test.rb]: We need to assert that the permitted flag is also kept when doing the deep dup.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by rafaelfranca** [comment]
Backported in bb3c364

**Pull request merged**